src/python/pants/core/util_rules/external_tool.py

```diff
@@ -128,15 +157,21 @@ def version(self) -> str:
     def known_versions(self) -> Tuple[str, ...]:
         return tuple(self.options.known_versions)
 
-    @abstractmethod
-    def generate_url(self, plat: Platform) -> str:
+    @property
+    def url_template(self) -> str:
+        return cast(str, self.options.url_template)
+
+    @property
+    def url_platform_mapping(self) -> Optional[Dict[str, str]]:
+        return cast(Optional[Dict[str, str]], self.options.url_platform_mapping)
+
+    def _generate_url(self, plat: Platform) -> str:
         """Returns the URL for the given version of the tool, runnable on the given os+arch.
 
         os and arch default to those of the current system.
-
-        Implementations should raise ExternalToolError if they cannot resolve the arguments
-        to a URL. The raised exception need not have a message - a sensible one will be generated.
         """
+        platform = self.url_platform_mapping[plat.value] if self.url_platform_mapping else ""
```

Comment:
Unclear to me how eagerly we should validate the `--url-platform-mapping`, e.g. ensure it aligns with the `Platform` enum.